osu.Game/Modes/Objects/HitObjectParser.cs

```diff
@@ -2,15 +2,68 @@
 //Licensed under the MIT Licence - https://raw.githubusercontent.com/ppy/osu/master/LICENCE
 
 using System;
-using System.Collections.Generic;
-using System.Linq;
-using System.Text;
-using System.Threading.Tasks;
+using System.Globalization;
+using osu.Game.Beatmaps;
+using osu.Game.Beatmaps.Samples;
 
 namespace osu.Game.Modes.Objects
 {
     public abstract class HitObjectParser
     {
-        public abstract HitObject Parse(string text);
+        public abstract HitObject Parse(Beatmap beatmap, string text);
+
+        public class UnknownHitObjectException : Exception
+        {
+            public int Type { get; }
+            public UnknownHitObjectException(int type)
+                : base($@"Unknown HitObject type {type}")
+            {
+                Type = type;
+            }
+        }
+
+        protected HitSampleInfo ParseHitSample(SampleInfo section, string sample, string addition)
+        {
+            HitSampleInfo hitSample = new HitSampleInfo
+            {
+                Type = (SampleType)Convert.ToInt32(sample, NumberFormatInfo.InvariantInfo) == SampleType.None ?
+                    SampleType.Normal :
+                    (SampleType)Convert.ToInt32(sample, NumberFormatInfo.InvariantInfo),
+                Bank = section.Bank,
+                Volume = section.Volume,
+                File = null
+            };
+
+            addition = string.IsNullOrWhiteSpace(addition) ? "0:0" : addition;
+            string[] split = addition.Split(':');
+            switch (split.Length)
+            {
+                case 5:
+                    hitSample.File = string.IsNullOrWhiteSpace(split[4]) ? null : split[4];
+                    goto case 4;

```

Comment:
rather than use `goto`, i usually either group the two cases together or create a second switch statement with the shared logic.